subsys/shell/shell.c

```diff
@@ -1302,19 +1287,15 @@ static int instance_uninit(const struct shell *sh)
 typedef void (*shell_signal_handler_t)(const struct shell *sh);
 
 static void shell_signal_handle(const struct shell *sh,
-				enum shell_signal sig_idx,
+				enum shell_signal sig,
 				shell_signal_handler_t handler)
 {
-	struct k_poll_signal *sig = &sh->ctx->signals[sig_idx];
-	int set;
-	int res;
-
-	k_poll_signal_check(sig, &set, &res);
-
-	if (set) {
-		k_poll_signal_reset(sig);
-		handler(sh);
+	if (!k_event_test(&sh->ctx->signal_event, sig)) {
+		return;
 	}
+
+	handler(sh);
+	k_event_clear(&sh->ctx->signal_event, sig);
```

Comment:
The order here changed vs how it used to work, not sure if possible but it could affect the behaviour. If, for example, the `handler` could re-signal the event?

Reply:
nice catch! restored order :)